File: scripts/extract-styles.mjs
#!/usr/bin/env node
/**
 * Extract StyleSheet.create blocks into separate .styles.ts files.
 * Usage: node scripts/extract-styles.mjs <file1> <file2> ...
 */

import { readFileSync, writeFileSync } from 'fs';
import { basename, extname } from 'path';

function findStylesBlock(content) {
  const pattern = /const\s+(\w+)\s*=\s*StyleSheet\.create\(/;
  const match = pattern.exec(content);
  if (!match) return null;

  const varName = match[1];
  const start = match.index;
  let i = match.index + match[0].length; // right after opening '('
  let depth = 1;
  let inStr = false;
  let strChar = null;

  while (i < content.length) {
    const c = content[i];
    if (inStr) {
      if (c === '\\' && strChar !== '`') { i += 2; continue; }
      if (c === strChar) inStr = false;
    } else {
      if (c === '"' || c === "'" || c === '`') { inStr = true; strChar = c; }
      else if (c === '(') depth++;
      else if (c === ')') {
        depth--;
        if (depth === 0) {
          let end = i + 1;
          if (content[end] === ';') end++;
          return { varName, block: content.slice(start, end), start, end };
        }
      }
    }
    i++;
  }
  return null;
}

function findModuleLevelVar(content, ident) {
  // Match: const SCREEN_HEIGHT = ... or const { height: SCREEN_HEIGHT } = ...
  const re = new RegExp(
    `(?:^|\\n)(const\\s+(?:\\{[^}]*\\b${ident}\\b[^}]*\\}|${ident})\\s*=\\s*[^\\n]+)`,
    ''
  );
  const m = re.exec(content);
  if (!m) return null;
  return m[1].trimEnd().replace(/;?$/, ';');
}

function getNeededImports(block, original) {
  const rnItems = ['StyleSheet'];
  if (/\bDimensions\b/.test(block)) rnItems.push('Dimensions');
  if (/\bPlatform\b/.test(block)) rnItems.push('Platform');

  // All UPPER_CASE identifiers in styles
  const allCaps = new Set([...block.matchAll(/\b([A-Z][A-Z0-9_]+)\b/g)].map(m => m[1]));
  for (const skip of ['StyleSheet', 'Platform', 'Dimensions', 'OS']) allCaps.delete(skip);

  // Which come from @/constants/colors?
  const colorsMatch = /import\s*\{([^}]+)\}\s*from\s*["']@\/constants\/colors["']/.exec(original);
  const colorImports = new Set();
  if (colorsMatch) {
    const available = new Set(colorsMatch[1].split(',').map(s => s.trim()).filter(Boolean));
    for (const c of allCaps) if (available.has(c)) colorImports.add(c);
  }

  // Remaining caps — check if they're module-level vars
  const moduleVars = {};
  for (const ident of allCaps) {
    if (colorImports.has(ident)) continue;
    const decl = findModuleLevelVar(original, ident);
    if (decl) moduleVars[ident] = decl;
  }

  return { rnItems, colorImports: [...colorImports].sort(), moduleVars };
}

function buildStylesFile(block, rnItems, colorImports, moduleVars) {
  const lines = [];
  lines.push(`import { ${rnItems.join(', ')} } from "react-native";`);
  if (colorImports.length)
    lines.push(`import { ${colorImports.join(', ')} } from "@/constants/colors";`);
  if (Object.keys(moduleVars).length) {
    lines.push('');
    for (const decl of Object.values(moduleVars)) lines.push(decl);
  }
  lines.push('');

  // Replace "const <varName> = " with "export default "
  const exportBlock = block.replace(/^const\s+\w+\s*=\s*/, 'export default ').replace(/;$/, '');
  lines.push(exportBlock + ';');
  lines.push('');

  return lines.join('\n');
}

function updateOriginal(content, varName, start, end, importPath) {
  // Remove styles block
  const before = content.slice(0, start).trimEnd();
  const after = content.slice(end).replace(/^\n+/, '');
  let out = before + '\n' + after;

  // Remove StyleSheet from react-native import if no longer needed elsewhere
  const rnImportRe = /import\s*\{([^}]+)\}\s*from\s*["']react-native["'];?\n?/;
  const stripped = out.replace(rnImportRe, '');
  if (!stripped.includes('StyleSheet')) {
    out = out.replace(rnImportRe, (_, group) => {
      const parts = group.split(',').map(s => s.trim()).filter(s => s && s !== 'StyleSheet');
      if (!parts.length) return '';
      return `import { ${parts.join(', ')} } from "react-native";\n`;
    });
    // Collapse triple+ blank lines
    out = out.replace(/\n{3,}/g, '\n\n');
  }

  // Add styles import after the last import line
  const importLines = [...out.matchAll(/^import\s.+/gm)];
  if (importLines.length) {
    const last = importLines[importLines.length - 1];
    const pos = last.index + last[0].length;
    out = out.slice(0, pos) + `\nimport ${varName} from "${importPath}";` + out.slice(pos);
  } else {
    out = `import ${varName} from "${importPath}";\n` + out;
  }

  return out;
}

const files = process.argv.slice(2);
for (const filepath of files) {
  try {
    const content = readFileSync(filepath, 'utf8');
    const found = findStylesBlock(content);
    if (!found) { console.log(`${filepath}: SKIP — no StyleSheet.create found`); continue; }

    const { varName, block, start, end } = found;
    const { rnItems, colorImports, moduleVars } = getNeededImports(block, content);

    const base = filepath.replace(/\.[^.]+$/, '');
    const stylesPath = base + '.styles.ts';
    const importPath = './' + basename(base) + '.styles';

    writeFileSync(stylesPath, buildStylesFile(block, rnItems, colorImports, moduleVars));
    writeFileSync(filepath, updateOriginal(content, varName, start, end, importPath));

    const extras = [];
    if (Object.keys(moduleVars).length) extras.push(`module_vars=[${Object.keys(moduleVars).join(',')}]`);
    if (varName !== 'styles') extras.push(`var=${varName}`);
    console.log(`${filepath}: OK → ${stylesPath}${extras.length ? ' [' + extras.join(', ') + ']' : ''}`);
  } catch (e) {
    console.error(`${filepath}: ERROR — ${e.message}`);
    console.error(e.stack);
  }
}

File: scripts/extract-styles.py
#!/usr/bin/env python3
"""
Extract StyleSheet.create blocks into separate .styles.ts files.
Usage: python3 scripts/extract-styles.py <file1> <file2> ...
"""

import re
import sys
import os


def find_styles_block(content):
    """Find StyleSheet.create block. Returns (var_name, block_text, start, end) or (None,...,-1,-1)."""
    pattern = re.compile(r'(const\s+(\w+)\s*=\s*StyleSheet\.create\()')
    match = pattern.search(content)
    if not match:
        return None, None, -1, -1

    var_name = match.group(2)
    start = match.start()
    i = match.end()  # right after the opening '('

    depth = 1  # we're inside the opening paren
    in_str = False
    str_char = None

    while i < len(content):
        c = content[i]

        if in_str:
            if c == '\\' and str_char != '`':
                i += 2
                continue
            if c == str_char:
                in_str = False
        else:
            if c in ('"', "'", '`'):
                in_str = True
                str_char = c
            elif c == '(':
                depth += 1
            elif c == ')':
                depth -= 1
                if depth == 0:
                    end = i + 1
                    if end < len(content) and content[end] == ';':
                        end += 1
                    return var_name, content[start:end], start, end
        i += 1

    return None, None, -1, -1


def find_module_level_vars(content, identifiers):
    """
    Find module-level const declarations for given identifiers (non-color, non-RN).
    Returns dict of {name: full_declaration_line}.
    """
    result = {}
    for ident in identifiers:
        # Match: const SCREEN_HEIGHT = ... or const { height: SCREEN_HEIGHT } = ...
        m = re.search(
            rf'(?m)^const\s+(?:\{{[^}}]*\b{re.escape(ident)}\b[^}}]*\}}\s*=\s*.+|{re.escape(ident)}\s*=\s*.+)',
            content
        )
        if m:
            result[ident] = m.group(0).rstrip(';') + ';'
    return result


def get_needed_imports(styles_block, original_content):
    """Determine what the styles file needs to import."""
    rn_items = ['StyleSheet']
    if re.search(r'\bDimensions\b', styles_block):
        rn_items.append('Dimensions')
    if re.search(r'\bPlatform\b', styles_block):
        rn_items.append('Platform')

    # Find all UPPER_CASE identifiers used in styles
    all_caps = set(re.findall(r'\b([A-Z][A-Z0-9_]+)\b', styles_block))
    all_caps -= {'StyleSheet', 'Platform', 'Dimensions', 'OS'}

    # Which of those come from @/constants/colors?
    colors_match = re.search(
        r'import\s*\{([^}]+)\}\s*from\s*["\']@/constants/colors["\']',
        original_content
    )
    color_imports = set()
    if colors_match:
        available = {c.strip() for c in colors_match.group(1).split(',') if c.strip()}
        color_imports = all_caps & available

    # Which are module-level vars (SCREEN_WIDTH, SCREEN_HEIGHT, etc.)?
    other_caps = all_caps - color_imports
    module_vars = find_module_level_vars(original_content, other_caps)

    return rn_items, sorted(color_imports), module_vars


def build_styles_file(var_name, styles_block, rn_items, color_imports, module_vars):
    lines = []
    lines.append(f'import {{ {", ".join(rn_items)} }} from "react-native";')
    if color_imports:
        lines.append(f'import {{ {", ".join(color_imports)} }} from "@/constants/colors";')
    if module_vars:
        lines.append('')
        for decl in module_vars.values():
            lines.append(decl)
    lines.append('')

    # Replace "const <var_name> = " with "export default "
    export_block = re.sub(r'^const\s+\w+\s*=\s*', 'export default ', styles_block)
    # Strip trailing semicolon after the closing ')' — export default doesn't need it
    export_block = export_block.rstrip(';').rstrip()
    lines.append(export_block + ';')
    lines.append('')

    return '\n'.join(lines)


def update_original(content, var_name, start, end, styles_import_path):
    """Remove styles block, clean up StyleSheet import, add styles import."""

    # --- Remove styles block ---
    before = content[:start].rstrip('\n')
    after = content[end:].lstrip('\n')
    new = before + '\n' + after

    # --- Remove StyleSheet from react-native import if no longer needed ---
    # Check if StyleSheet is used anywhere else in the file (outside the block we removed)
    remaining = new
    # Strip the RN import line temporarily to check residual usage
    stripped = re.sub(r'import\s*\{[^}]+\}\s*from\s*["\']react-native["\'];?\n?', '', remaining, count=1)
    if 'StyleSheet' not in stripped:
        def _remove_stylesheet(m):
            parts = [p.strip() for p in m.group(1).split(',')]
            parts = [p for p in parts if p and p != 'StyleSheet']
            if not parts:
                return ''
            return 'import { ' + ', '.join(parts) + ' } from "react-native";'
        new = re.sub(
            r'import\s*\{([^}]+)\}\s*from\s*["\']react-native["\'];?',
            _remove_stylesheet,
            new,
            count=1
        )
        # Clean up blank lines left by a completely removed import
        new = re.sub(r'\n{3,}', '\n\n', new)

    # --- Add styles import after last import statement ---
    last_import = None
    for m in re.finditer(r'^import\s.+', new, re.MULTILINE):
        last_import = m
    if last_import:
        pos = last_import.end()
        new = new[:pos] + f'\nimport {var_name} from "{styles_import_path}";' + new[pos:]
    else:
        new = f'import {var_name} from "{styles_import_path}";\n' + new

    return new


def process(filepath):
    with open(filepath, 'r') as f:
        content = f.read()

    var_name, block, start, end = find_styles_block(content)
    if var_name is None:
        return 'SKIP — no StyleSheet.create found'

    rn_items, color_imports, module_vars = get_needed_imports(block, content)

    base = os.path.splitext(filepath)[0]
    styles_path = base + '.styles.ts'
    basename = os.path.basename(base)
    import_path = f'./{basename}.styles'

    styles_src = build_styles_file(var_name, block, rn_items, color_imports, module_vars)
    with open(styles_path, 'w') as f:
        f.write(styles_src)

    updated = update_original(content, var_name, start, end, import_path)
    with open(filepath, 'w') as f:
        f.write(updated)

    extras = []
    if module_vars:
        extras.append(f'module_vars={list(module_vars.keys())}')
    if var_name != 'styles':
        extras.append(f'var={var_name}')
    suffix = f' [{", ".join(extras)}]' if extras else ''
    return f'OK → {styles_path}{suffix}'


if __name__ == '__main__':
    for filepath in sys.argv[1:]:
        try:
            print(f'{filepath}: {process(filepath)}')
        except Exception as e:
            import traceback
            print(f'{filepath}: ERROR — {e}')
            traceback.print_exc()
